### PR TITLE
ENH: add `entropy` method for multivariate t distribution

### PR DESCRIPTION
**Pull request by dschmitz89**
#### What does this implement/fix?
Adds an entropy method for the multivariate t distribution.

#### Additional information
Source: https://math.stackexchange.com/questions/2272184/differential-entropy-of-the-multivariate-student-t-distribution

image

One quick (and not necessarily sufficient) confirmation that the formula is correct_ for $df\to\infty$, we get the entropy of the multivariate normal.

image

<details>

```
from scipy import stats
from scipy import special as sc
import numpy as np
import matplotlib.pyplot as plt

def multivariate_t_entropy_analytic(df, cov):
    cov = np.atleast_1d(cov)
    dim = cov.shape[0]
    halfsum = 0.5 * (dim + df)
    half_df = 0.5 * df
    _, logdet = np.linalg.slogdet(cov)
    return (-sc.gammaln(halfsum) + sc.gammaln(half_df)
            + 0.5 * dim * np.log(df * np.pi)
            + halfsum * (sc.psi(halfsum) - sc.psi(half_df))
            + 0.5 * logdet)

dfs = np.logspace(1, 5)
cov = np.eye(5, 5)
entropy = multivariate_t_entropy_analytic(dfs, cov)
plt.semilogx(dfs, entropy, label="t distribution", ls="dashdot")
plt.axhline(stats.multivariate_normal.entropy(np.zeros((cov.shape[0], )), cov=cov), 0, 1,
            label="normal distribution", c='r', ls="dashed")
plt.xlabel("$df$")
plt.ylabel("Entropy")
plt.legend()
plt.title("Differential entropies: $\Sigma=I_5$")
```

</details>

### CODE REVIEW

**Comment by mdhaber** [comment]
Looks like you pushed a submodule commit. I can fix it if you need.
I'd be OK with computing the integral numerically in the test (rather than in a comment) if it's not too slow.
@tirthasheshpatel are you willing to review this?

**Comment by dschmitz89** [comment]
> Looks like you pushed a submodule commit. I can fix it if you need.

Those nasty submodules .. could not find a quick fix. If you know an easy trick apart from reverting and force pushing, please go ahead :).

> I'd be OK with computing the integral numerically in the test (rather just including the code in a comment) if it's not too slow. 

They take around 30 seconds per case on my machine, I guess that is too much. I thought a reviewer might even ask for more cases in other dimensions. Interestingly, I cannot get useful results from mpmath'S `quad` here (can provide if interested).

**Comment by mdhaber** [comment]
If what I suggested above doesn't work, you can probably do the same as gh-18304. If that doesn't work, I'll look into it.

**Comment by mdhaber** [comment]
I'm assuming that this branch was created from `main` after the submodule commit was updated but before you had checked out the latest commit locally. `git status` would have shown something like this:

```
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   scipy/_lib/boost_math (new commits)
``
in addition to your other changes because it sees that the submodule commit you have checked out doesn't match what it's supposed to be. Perhaps you didn't notice, and just did `git add scipy` or similar, and then you committed.

If that was the state of things, the simplest way to have fixed it would be to:
- `git checkout main`
- `git pull` (or fetch from upstream and merge)
- `git submodule update --init` (to checkout the right submodule commit)
- `git checkout multivariate_t_entropy`
- `git status` should show that there are changes to the submodule
- `git add scipy`
- `git commit`

File: scipy/stats/_multivariate.py [hunk 1]
```diff
@@ -4550,6 +4556,35 @@ def cdf(self, x, loc=None, shape=1, df=1, allow_singular=False, *,
...
+                * (psi(halfsum) - psi(half_df))
+                + 0.5 * shape_info.log_pdet)
+
+    def entropy(self, shape=1, df=1):
+        """Calculate the differential entropy of a multivariate
+        t-distribution.
+
+        Parameters
+        ----------
+        %(_mvt_doc_default_callparams)s
```
Comment (tirthasheshpatel) [inline scipy/stats/_multivariate.py:4577]: This is never expanded. Need to add `cdf` to the list at line 4770.

**Review by tirthasheshpatel — APPROVED** [review]
Thanks for the changes @dschmitz89! Looks good to me now! @mdhaber If you don't have any comments, we can merge!

**Comment by mdhaber** [comment]
Not a full review, but a few thoughts:
- multivariate normal entropy (and all univariate distribution entropy methods) accept a location parameter even though they don't use it. For consistency, I'd expect multivariate t entropy to accept it, too.
- The tests tests would be a little more convincing to me with a shape matrix other than the identity.
- I'm surprised that the numerical integration doesn't get closer agreement in 2D with such a well-behaved covariance. Do the tails have a big effect, maybe because of the low `df`? (How does it do with smaller magnitude `limit`)?

**Comment by dschmitz89** [comment]
> Not a full review, but a few thoughts:
> 
> * multivariate normal entropy (and all univariate distribution entropy methods) accept a location parameter even though they don't use it. For consistency, I'd expect multivariate t entropy to accept it, too.

That makes sense for consistency, will do.

> * The tests would be a little more convincing to me with a shape matrix other than the identity. Is the formula appropriate for singular shape, too, for example?

Will test it out.

> * I'm surprised that the numerical integration doesn't get closer agreement in 2D with such a well-behaved covariance. Do the tails have a big effect, maybe because of the low `df`? (How does it do with smaller magnitude `limit`)?

The numerical integration is in general not a perfect ground truth here. For the 2D test examples it returns 0 for $df>2$ already. The tolerance can be lowered to `1e-5` though.

**Comment by mdhaber** [comment]
There are two things that are making it difficult to compare:
- With such low `df` the tails make a big difference
- The `limit` is just too high.

Please increase `df` and reduce `limit` in the tests.

**Comment by tirthasheshpatel** [comment]
> do you agree?

I agree. Would be nice to get the reference values more accurate. Also, IIUC, the current implementation should support singular matrices, right @dschmitz89? If yes, would be nice to add a couple of tests for such matrices too. If you want to do that as a follow-up, fine with me!

**Review by tirthasheshpatel — APPROVED** [review]
Thanks @dschmitz89! Also, thanks @mdhaber for the review! The PR looks good now. Just one point remaining: adding tests for singular matrices. Let's take that up in a follow-up. Merging!

**Pull request merged**